src/pybind/mgr/mds_autoscaler/module.py

```diff
@@ -0,0 +1,152 @@
+"""
+
+import logging
+from typing import Optional, List
+from mgr_module import MgrModule
+from ceph.deployment.service_spec import ServiceSpec
+import orchestrator
+import copy
+
+log = logging.getLogger(__name__)
+
+
+class MDSAutoscaler(orchestrator.OrchestratorClientMixin, MgrModule):
+    """
+    MDS autoscaler.
+    """
+    def __init__(self, *args, **kwargs):
+        super(MDSAutoscaler, self).__init__(*args, **kwargs)
+        self.set_mgr(self)
+        self.fs_map: Optional[dict] = None
+
+    def get_service(self, fs_name: str) -> Optional[List[orchestrator.ServiceDescription]]:
+        result = None
+        try:
+            service = "mds.{}".format(fs_name)
+            completion = self.describe_service(service_type='mds',
+                                               service_name=service,
+                                               refresh=True)
+            self._orchestrator_wait([completion])
+            orchestrator.raise_if_exception(completion)
+            result = completion.result
+        except Exception as e:
+            self.log.exception("{}: exception while fetching service info"
+                               .format(e))
+            pass
+        return result
+
+    def get_daemons(self, fs_name: str) -> Optional[List[orchestrator.DaemonDescription]]:
+        result = None
+        try:
+            service = "mds.{}".format(fs_name)
+            completion = self.list_daemons(service_name=service)
+            self._orchestrator_wait([completion])
+            orchestrator.raise_if_exception(completion)
+            result = completion.result
+        except Exception as e:
+            self.log.exception("{}: exception while fetching service info"
+                               .format(e))
+            pass
+        return result
+
+    def update_daemon_count(self, fs_name: str, abscount: int) -> Optional[ServiceSpec]:
+        svclist = self.get_service(fs_name)
+        if svclist is None or len(svclist) == 0:
+            self.log.warn("Failed to fetch MDS service details for fs '{}'"
+                          .format(fs_name))
+            return None
+
+        svc = svclist[0]
+
+        # no need to update placement count if the current placement count is
+        # the required count
+        if svc.spec.placement.count == abscount:
+            return None
+
+        ps = copy.deepcopy(svc.spec.placement)
+        ps.count = abscount
+        newspec = ServiceSpec(service_type=svc.spec.service_type,
+                              service_id=svc.spec.service_id,
+                              placement=ps)
+        return newspec
+
+    def get_required_standby_count(self, fs_name: str) -> Optional[int]:
+        total = 0
+        assert self.fs_map is not None
+        for fs in self.fs_map['filesystems']:
+            if fs['mdsmap']['fs_name'] == fs_name:
+                return fs['mdsmap'].get('standby_count_wanted')
+        return total
```

Comment:
This line should never be hit. Is this just defensive coding?

Reply:
this is indeed defensive coding
we could have an assert here instead ... but either way the system is going to assert and fail or cause standby mds count to drop if the fs_map is not good